becquerel/core/peakfinder.py

```diff
@@ -0,0 +1,395 @@
+# -*- coding: utf-8 -*-
+
+"""Spectral peak search using convolutions."""
+
+from __future__ import print_function
+import matplotlib.pyplot as plt
+import numpy as np
+from .spectrum import Spectrum
+
+
+class PeakFilterError(Exception):
+    """Base class for errors in PeakFilter."""
+
+    pass
+
+
+class PeakFilter(object):
+    """An energy-dependent kernel that can be convolved with a spectrum.
+
+    To detect lines, a kernel should have a positive component in the center
+    and negative wings to subtract the continuum, e.g., a boxcar:
+
+    +2|     ┌───┐     |
+      |     │   │     |
+     0|─┬───┼───┼───┬─|
+    -1| └───┘   └───┘ |
+
+    The positive part is meant to detect the peak, and the negative part to
+    sample the continuum under the peak.
+
+    The kernel should sum to 0.
+
+    The width of the kernel scales proportionally to the square root
+    of the channels, with a minimum value set by 'fwhm_at_0'.
+
+    """
+
+    def __init__(self, ref_channel, ref_fwhm, fwhm_at_0=1.):
+        """Initialize with a reference line position and FWHM in channels."""
+        if ref_channel <= 0:
+            raise PeakFilterError('Reference channel must be positive')
+        if ref_fwhm <= 0:
+            raise PeakFilterError('Reference FWHM must be positive')
+        if fwhm_at_0 < 0:
+            raise PeakFilterError('FWHM at 0 must be non-negative')
+        self.ref_channel = float(ref_channel)
+        self.ref_fwhm = float(ref_fwhm)
+        self.fwhm_at_0 = float(fwhm_at_0)
+
+    def fwhm(self, channel):
+        """Calculate the expected FWHM at the given channel."""
+        # f(x)^2 = f0^2 + k x^2
+        # f1^2 = f0^2 + k x1^2
+        # k = (f1^2 - f0^2) / x1^2
+        # f(x)^2 = f0^2 + (f1^2 - f0^2) (x/x1)^2
+        f0 = self.fwhm_at_0
+        f1 = self.ref_fwhm
+        x1 = self.ref_channel
+        fwhm_sqr = f0**2 + (f1**2 - f0**2) * (channel / x1)**2
+        return np.sqrt(fwhm_sqr)
+
+    def kernel(self, channel, n_channels):
+        """Generate the kernel for the given channel."""
+        raise NotImplementedError
+
+    def kernel_matrix(self, n_channels):
+        """Build a matrix of the kernel evaluated at each channel."""
+        kern = np.zeros((n_channels, n_channels))
+        for j in range(n_channels):
+            kern[:, j] = self.kernel(j, n_channels)
+        kern_pos = +1 * kern.clip(0, np.inf)
```

Comment:
do we want to do, e.g. `kern.clip(0, np.inf)  # * (+1)`, so we don't have to do an extra unnecessary multiplication operation?

Reply:
The multiplication by +1 is for parallelism with the line below.